### DISCUSSION TITLE
V3.08d Released

### DISCUSSION THREAD

**Discussion by GerryDazoo** [GerryDazoo/Slinger#203] [Announcements]
This version now supports changing channels on the ProHD built in tuner. Grab the ProHD remote file in CustomRemotes. Tested with SMPlayer and FFPLayer. VLC is a little problematic when it sees the new video stream on the channel change. We found that this command line parameter makes it sort of usable “--no-play-and-pause”. There may be some other VLC configuration(s) to make it better but I’m going to leave that as an exercise for the user.

I revisited how IR keycodes were queued up for sending to the slingbox due to some user issues with the sending being “too slow”. Channel changing should now be much faster with fewer video artifacts.

The code can now eat a raw HTML file for the remote definition. Make your fancy remote page with your favourite web page editor and give it to the server with the “.html” file extension. All the buttons need a unique name= with the keycode in the value= parameter. To send a channel change the name of the input field needs to be “Digits”.

The code will now work with older versions of Python. Tested on 3.7

The server will now “logout” when a session with the slingbox ends. You should see fewer “kicked off” messages. This was primarily as issue on the Solo boxes.

You can now send keycodes when specifying the start channel either in the config.ini file or on the streaming URL . Both of these are supported.
?channel=155 #The channel number to tune to
?channel=9+13+13 # The keycodes

**Comment by DoogieHowitzerWMD** [comment]
I've been experiencing getting "kicked off" since the 3.08 update with the slingbox 500. I've been using VLC, and the stream will just stop suddenly, which hadn't been the case before. I have not yet checked the logs on the cause yet though (running on a headless RPi with ssh only, so I have to figure out how to open the logs from the command line).

**Reply by DoogieHowitzerWMD** [reply]
> I've been experiencing getting "kicked off" since the 3.08 update with the slingbox 500. I've been using VLC, and the stream will just stop suddenly, which hadn't been the case before. I have not yet checked the logs on the cause yet though (running on a headless RPi with ssh only, so I have to figure out how to open the logs from the command line).

I've determined my issue seems to be with either the slingbox itself or the network it is on. I'd say about half the time, the Slingbox will not respond to initial connection requests from the server, but when it does and it starts a stream, the stream will eventually disconnect within at least 5 minutes.

**Comment by bradut** [comment]
Thank you Gerry for this new version (3.09).
When running the server I am having an issue: 
```
PING failure (...) [WinError 10060] A connection attempt failed because the connected party did not properly respond after a period of time, or established connection failed because connected host has failed to respond
No working external portmaps. Not registering slingboxes
Retry 1 in 10 minutes
```
I was able to run the new server only one time, after resetting the SlingBox.
The next time I tried, got the same error as above.

The config.ini is the same as in previous version (3.08.c) and version 3.08.c works fine

**Reply by GerryDazoo** [reply]
Either turn off your FInderID or fix the portmap on your router. It's a background process that is running.
The code is still running but trying to access your server via my web server will not work.

**Reply by bradut** [reply]
I commented-out the FinderID and now the server stops after displaying  `* Debug mode: off`
The port mapping is the same, and works fine with 3.08.c.

**Reply by GerryDazoo** [reply]
Please post the server output. What you're seeing is the expected output until you try a connection from a media player.. Are you saying you can't start a video stream and see no server output?
Also try starting a stream using your external IP address from inside your LAN. That's basically what my code is complaining about. If you can do that and my code is still complaining then their might be a bug there.

**Reply by bradut** [reply]
- I am connecting with VLC and Remote from inside my network only by using my external IP & port.

- To check if someone is listening to the external forwarded port, I have used this port forward tester:  https://www.yougetsignal.com/tools/open-ports/
  - With 3.08.c it displays Port (my external port) is open on (my external IP)
  - With 3.09 it displays '... is closed...'
  
 - Server output. The server display this info and stops:
```
Version : 3.09 Running on Windows-10-10.0.19044-SP0 pid= 18596
Using config file config.ini
Connection Manager Running on port YYYY with 10 max streams using URL slingbox.
Streamer Running:  10 config.ini SLINGBOX YYYY YYYY  8388608
starting up on  port YYYY 8388608
Checking for slingbox at ('12.34.56.78', XXXX)
('12.34.56.78', XXXX) OK
Bradut TV Using slingbox at  ('12.34.56.78', XXXX)
Streamer:  Bradut TV Waiting for first stream, flushing any IR requests that arrive while not connected to slingbox
Reading Custom Remote definition from remote.txt
 * Serving Flask app 'slingbox_server' (lazy loading)
 * Environment: production
   WARNING: This is a development server. Do not use it in a production deployment.
   Use a production WSGI server instead.
 * Debug mode: off  
```
Then, after a while, it displays the error msg I've mentioned above:
```
10/31/2022, 18:11:11.425  Registering Slingboxes {'4A62CBE24FB68440B6024FA97C26FBB7': XXXX}
Getting external IP address
Redirector, my external IP address 12.34.56.78
10/31/2022, 18:11:32.777  PING failure ('12.34.56.78', XXXX) [WinError 10060] A connection attempt failed because the connected party did not properly respond after a period of time, or established connection failed because connected host has failed to respond
No working external portmaps. Not registering slingboxes
Retry 1 in 10 minutes
10/31/2022, 18:21:32.791  Registering Slingboxes {'4A62CBE24FB68440B6024FA97C26FBB7': XXXX}
Getting external IP address
Redirector, my external IP address 12.34.56.78
10/31/2022, 18:21:54.031  PING failure ('12.34.56.78', XXXX) [WinError 10060] A connection attempt failed because the connected party did not properly respond after a period of time, or established connection failed because connected host has failed to respond
No working external portmaps. Not registering slingboxes
Retry 2 in 10 minutes
```

- Given that the server is not listening to the assigned port, VLC and Remote are not working

**Reply by GerryDazoo** [reply]
The server should be listening, that is basic code that hasn't changed in a long time. The background process trying to verify you have a proper  port map in your router to allow external access is failing, it will try for a while then give up.
Please get the new exe file from github, It should say V3.08d
In a command shell please run a "netstat -an" command. you should see a TCP socket listening on port XXXX
If you run as administrator netstat -anb you should see this
  TCP    0.0.0.0:XXXX          0.0.0.0:0              LISTENING
 [slingbox_server.exe]

  TCP    127.0.0.1:XXXX+1        0.0.0.0:0              LISTENING
 [slingbox_server.exe]
Off-topic, Is there a reason why you're using your external IP address to access your slingbox when it is on the same  local LAN as your server?

**Reply by bradut** [reply]
I just gave it a try. It says "Version : 3.08d" as you indicated, but still fails with the same error. 
Will come back with more details in the morning, as it's 1 AM here, but couldn't fall sleep after reading your message about the new version :)

**Reply by GerryDazoo** [reply]
OK. After a few hours of sleep. I have a pretty good idea what might be wrong. In the readme there was some discussion about some anti-virus software (Windows Defender) thinking that there was a virus in the executable and blocking it. This is what I think is happening to you, the new executable is being blocked from accessing the network. Please disable and anti-virus software temporarily to verify.
Note this will be more apparent after to run the netstat command and don't see any executable listening on port XXXX.

**Reply by bradut** [reply]
Indeed, **your ideea was great** !
I added the new executable to the firewall and it works !
Thank you !

Regarding the reason why I am using your external IP address to access my SlingBox when it is on the same local LAN as my server:
- I just want to make sure that I can access my SlingBox from outside. Eventually I will switch to internal IPs only, to reduce traffic counted by my ISP.

BTW: When I try to access the sling server from another network:
- it works 
- the server displays a warning:
```
Starting remote stream 3
Bradut TV New Stream Starting 0
11/01/2022, 16:12:24.509  Bradut TV Stream Terminated for  12.34.56.78:37060 [WinError 10035] A non-blocking socket operation could not be completed immediately
```
Despite the warning above, I can stream in VLC.

**Comment by theweatherguy** [comment]
Gerry,
I'd like to make a custom remote for my FIOS Verizon remote.  In your 3.08d release notes, you indicate:

_"The code can now eat a raw HTML file for the remote definition. Make your fancy remote page with your favourite web page editor and give it to the server with the “.html” file extension."_

> What file do you create with an .html extension?   Where do you place it so the code can 'eat it' .    I'm lost with this.  Does anyone have a specific example of a remote .html file and explain how to use it?

_"All the buttons need a unique name= with the keycode in the value= parameter. To send a channel change the name of the input field needs to be “Digits”._ 

> I guess if I will be able to figure this part out after I figure out the earlier question on the remote html file.

**Reply by GerryDazoo** [reply]
If you don't need anything fancy then making your own remote.html file is probably not for you. There is already a custom FIOS_remote.txt file available in custom remotes  which would be a good starting point. 

So, if you don't have a "favorite web page editor" best to pass. But to answer your questions. 
1) You create an HTML page with an ,html file extension using your "favorite web page editor"
2) You replace your remote_def.txt file with remote_def.html in the config.ini file.

No one has built an html remote file yet. I just tested my code with the html generated for remote.txt. In firefox, open a remote connection and right click. Select view page source to see what the html looks like.

**Reply by theweatherguy** [reply]
Ok, so the custom .html file is referenced and pointed to through the name in the remote section of the config.ini file.  And the remote html file would reside as a separate file in the same directory as the slingbox_server.py and config.ini files. (?)

If no one has built that yet, I'm not going to be the first.  LOL

**Reply by GerryDazoo** [reply]
Yes, in the [REMOTE] section or the remote entry in the unified config file.
Note: The remote file can be anywhere. It does not have to be in the same folder as the slingbox_server code.
Just use the full path to it in your config file
IE. C:\some_folder\some_other_subfolder\remote_file_name

**Reply by lepmail** [reply]
The Weather Guy -- take a look at what I uploaded [here](https://github.com/GerryDazoo/Slinger/discussions/224).

**Comment by Pthepeg** [comment]
Hi Gerry, 3.08d isn't finding the slingbox on my setup, the message is different this time. Note that if I define the slingbox IP address in the config file, all is okay. Python 3.7.3. I can try a patch if you send one.
-----------------------------
$ python3 slingbox_server.py configunified320.ini
Version : 3.08d Running on Linux-5.10.103+-armv6l-with-debian-10.13 pid= 6028
Using config file configunified320.ini
Connection Manager Running on port 62433 with 2 max streams using URL stream.
BOXES [('sb1', 'slingJim')]
Streamer Running:  2 configunified320.ini slingJim slingJim 62433 8388608
starting up on  port 62433 8388608
slingJim No valid slingbox ip info found in config.ini
Finding Slingbox on local network. My IP Info =  b8:27:eb:97:99:75
Error binding socket to send broadcast [Errno -2] Name or service not known
Can't find a slingbox on network. Please make sure it's plugged in and connected. Check config.ini
Building page for slingJim
Reading Custom Remote definition from /home/slingbox/remote.txt
 * Serving Flask app "slingbox_server" (lazy loading)
 * Environment: production
   WARNING: Do not use the development server in a production environment.
   Use a production WSGI server instead.
 * Debug mode: off
slingJim Giving up. Sorry...

**Reply by GerryDazoo** [reply]
replace line 126
with the word continue
This was some debugging code I forgot to delete. Sorry.
But I am curious what IPV4 interface on your Pi has a MAC address instead of an IP address.
run this command > ip a
Unfortunately I don't have my Pi on hand to test with. In the future would you be willing to take new code out for a test drive?

**Reply by Pthepeg** [reply]
line 126 patch fixes the problem. thanks.

I don't know why its a MAC address, have a look:
$ ip a
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host 
       valid_lft forever preferred_lft forever
2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast state UP group default qlen 1000
    link/ether b8:27:eb:97:99:75 brd ff:ff:ff:ff:ff:ff
    inet 192.168.100.127/24 brd 192.168.100.255 scope global dynamic noprefixroute eth0
       valid_lft 43795sec preferred_lft 32995sec
    inet6 fe80::98ae:2acf:dd39:1e71/64 scope link 
       valid_lft forever preferred_lft forever

**Reply by GerryDazoo** [reply]
Interesting why my code found two interfaces but the same MAC and IP address were assigned to eth0. I'll have a look at the code. 
Grab the latest 3.08e.

**Comment by rblancarte** [comment]
Thank you for this update! This appears to have resolved my issues with the slow remote trouble I was having.
